UI/window-dock-youtube-app.cpp

```diff
@@ -0,0 +1,482 @@
+static constexpr const char *BROADCAST_SELECTED = "BROADCAST_SELECTED";
+static constexpr const char *INGESTION_STARTED = "INGESTION_STARTED";
+static constexpr const char *INGESTION_STOPPED = "INGESTION_STOPPED";
+
+YouTubeAppDock::YouTubeAppDock()
+	: BrowserDock(),
+	  dockBrowser(nullptr),
+	  cookieManager(nullptr)
+{
+	OBSBasic::InitBrowserPanelSafeBlock();
+	AddYouTubeAppDock("YouTube Live Control Panel");
+}
+
+YouTubeAppDock::~YouTubeAppDock()
+{
+	if (cookieManager) {
+		cookieManager->FlushStore();
+		delete cookieManager;
+	}
+}
+
+bool YouTubeAppDock::IsYTServiceSelected()
+{
+	obs_service_t *service_obj = OBSBasic::Get()->GetService();
+	OBSDataAutoRelease settings = obs_service_get_settings(service_obj);
+	const char *service = obs_data_get_string(settings, "service");
+	return IsYouTubeService(service);
+}
+
+void YouTubeAppDock::AccountConnected()
+{
+	channelId.clear(); // renew channel id
+	UpdateChannelId();
+}
+
+void YouTubeAppDock::AccountDisconnected()
+{
+	SettingsUpdated(true);
+}
+
+void YouTubeAppDock::SettingsUpdated(bool cleanup)
+{
+	bool ytservice = IsYTServiceSelected();
+	SetVisibleYTAppDockInMenu(ytservice);
+
+	// definitely cleanup if YT switched off
+	if (!ytservice || cleanup) {
+		if (cookieManager)
+			cookieManager->DeleteCookies("", "");
+	}
+	if (ytservice)
+		Update();
+}
+
+std::string YouTubeAppDock::InitYTUserUrl()
+{
+	std::string user_url(YOUTUBE_WEBAPP_PLACEHOLDER_URL);
+
+	if (IsUserSignedIntoYT()) {
+		YoutubeApiWrappers *apiYouTube = GetYTApi();
+		if (apiYouTube) {
+			ChannelDescription channel_description;
+			if (apiYouTube->GetChannelDescription(
+				    channel_description)) {
+				QString url =
+					QString(YOUTUBE_WEBAPP_ADDRESS_URL)
+						.arg(channel_description.id);
+				user_url = url.toStdString();
+			} else {
+				blog(LOG_ERROR,
+				     "YT: InitYTUserUrl() Failed to get channel id");
+			}
+		}
+	} else {
+		blog(LOG_ERROR, "YT: InitYTUserUrl() User is not signed");
+	}
+
+	blog(LOG_DEBUG, "YT: InitYTUserUrl() User url: %s", user_url.c_str());
+	return user_url;
+}
+
+void YouTubeAppDock::AddYouTubeAppDock(const QString &title)
+{
+	QString bId(QUuid::createUuid().toString());
+	bId.replace(QRegularExpression("[{}-]"), "");
+	this->setProperty("uuid", bId);
+	this->setObjectName(title + "Object");
+	this->resize(580, 500);
+	this->setMinimumSize(400, 300);
+	this->setWindowTitle(title);
+	this->setAllowedAreas(Qt::AllDockWidgetAreas);
+
+	OBSBasic::Get()->addDockWidget(Qt::RightDockWidgetArea, this);
+	actionYTAppDock = OBSBasic::Get()->AddDockWidget(this);
+
+	if (IsYTServiceSelected()) {
+		const std::string url = InitYTUserUrl();
+		CreateBrowserWidget(url);
+
+		// reload panel layout
+		const char *dockStateStr = config_get_string(
+			App()->GlobalConfig(), "BasicWindow", "DockState");
+		if (dockStateStr) {
+			QByteArray dockState = QByteArray::fromBase64(
+				QByteArray(dockStateStr));
+			OBSBasic::Get()->restoreState(dockState);
```

Comment:
This is bad, a dock type should not touch a dockState that is not the one used by the integration.